### ISSUE TITLE
nbconvert vs. interactive Jupyter operation

### ISSUE THREAD

**Issue by ElectricGizmo** [jupyter/nbconvert#659]
I am running the same notebook interactively and using nbconvert on Windows, clean Python 3 kernel installation (using Anaconda).

The notebook works just fine from within Jupyter but I get the error when converting the notebook to html to automate generation of event reports for colleagues. When I convert the notebook to html I get  
`[NbConvertApp] WARNING | Timeout waiting for IOPub output`  
and the output is truncated for one of the cells, not always in the exact same position (for the same input data), never with any exception detail.  
I switched the _raise_on_iopub_timeout_ flag and apparently `execute.py line 348 in run_cell` raised `RuntimeError("Timeout waiting for IOPup output")` and announces `The system cannot find the file specified`  
The notebook makes heavy (heavy!) use of matplotlib to show the data and crunches for several minutes in every case, whether it works (Jupyter) or not (nbconvert).  
Also worth mentioning, the graph generation is done in object methods defined in an external set of scripts first run by the notebook (no problem there) then invoked at the data display time.  
Again, it all works fine when in an interactive Jupyter-served page.  

- Is there a way to display the callstack to identify what file is not found? I suspect it is a png generated by matplotlib, but which one?
- What difference between interactive and nbconvert processing could explain the issue?

Thanks!

For the records I invoke nbconvert from a Python script as follows:
> os.system(r"jupyter nbconvert " +
>           r"--ExecutePreprocessor.timeout=None "+
>           r"--ExecutePreprocessor.iopub_timeout=10 " +
>           r"--to html " +
>           # r"--allow-errors " +
>           r"--ExecutePreprocessor.raise_on_iopub_timeout=True  " +
>           r"--execute {}ARManalysis.ipynb".format(localPath))

**Comment by takluyver** [comment]
An IOPub timeout means it has produced no output for a certain amount of time. You're setting that time to 10 seconds. Is it possible that your crunching code produces no output for 10 seconds.

I don't know where the "cannot find the file" error is coming from, but i suspect that's a red herring.

**Comment by mpacer** [comment]
Would it be possible to trigger something like the `%debug` magic that will drop you into a debugger if an error is raised when running --execute from the command line?

**Comment by ElectricGizmo** [comment]
@takluyver , thanks for your answer.
I would be surprised the processing takes more than twice longer when using nbconvert than when running from within Jupyter. Besides I tried with 600s with the same effect (except it took longer to announce the issue).
So I really believe it is a behavioral difference, not just timing.

**Comment by ElectricGizmo** [comment]
Well, thanks for the tip but even though I added %debug at the beginning of the cell where the problem occurs I ended-up empty-handed: when running the notebook interactively I get a red bx at the bottom of that cell saying "ERROR:root:No traceback has been produced, nothing to debug."
When I feed the same notebook to nbconvert I get the same IOPub timeout and a short cell output, no red box, no debug data.
Hmmmmm

**Comment by ElectricGizmo** [comment]
Whoops, I was looking at the bottom of the quite extensive output instead of the bottom of the code:
while the output is cut short, there IS a red box cell saying "ERROR:root:No traceback has been produced, nothing to debug."

**Comment by djherbis** [comment]
I did some digging with this and I'm pretty confident I know what's wrong with this issue.

**tl;dr; IOPub channel ZMQ buffer fills up, causing some output and the 'idle' message to be dropped, triggering an IOPub timeout waiting for the dropped msg**

**Why does it work interactively but not in NBConvert?**

In interactive mode, the Jupyter notebook consumes the IOPub messages as they are produced and displays them as live output. In NBConvert, it actually [waits until the cell has finished executing](https://github.com/jupyter/nbconvert/blob/1a7b848cedae10970f6df83268ef0898ba80d734/nbconvert/preprocessors/execute.py#L447) before consuming any of the output (causing it to buffer up).

You'll notice in the link above that the error message you get shows up right there. It's triggered because the buffered output of your cell fills up the IOPub Channel ZMQ buffer and prevents it from receiving any more output, so it just drops any more output sent. The IOPub channel is used both for sending output content, and for [sending control signals like 'idle'](https://github.com/jupyter/nbconvert/blob/1a7b848cedae10970f6df83268ef0898ba80d734/nbconvert/preprocessors/execute.py#L447) which signals that it's done. Since the idle message is dropped when the buffer is full, you'll get an IOPub timeout as it waits for this message.

A minimal python example which can trigger this:
```python
import sys
import time
str = '0'
for x in range(0, 10000):
  sys.stdout.write(str*100)
  sys.stdout.flush()
  time.sleep(0.0001)
```

Basically we produce a lot of output messages, while not exceeding the IOPub data rate which fills up the buffer and causes the idle message to be dropped.

**Can we fix it?**
Well there's a few options that I can think of, but I'm happy to hear suggestions from Jupyter contributors. 

1. Consume the output in nbconvert as it is sent instead of blocking for the cell to finish. In my tests just moving [this line](https://github.com/jupyter/nbconvert/blob/1a7b848cedae10970f6df83268ef0898ba80d734/nbconvert/preprocessors/execute.py#L447) made my example work, though there's certainly more to it than that (fixing timeouts, not spin looping checking for start of output etc.), probably can take hints from how the interactive client consumes the messages.

2. Use a separate channel for sending things like the 'idle' message. This would prevent the idle message from being dropped from too much output, but would still cause output to be dropped/cut-off.

3. Increase buffer sizes (only really defers the problem, but solves it if you stay under the buffer size). I tried adding `sock.setsockopt(zmq.RCVBUF, 20000000)` near [this line](https://github.com/jupyter/jupyter_client/blob/master/jupyter_client/connect.py#L545), and that also solved the issue. Presumably just tuning your kernels tcp buffer size directly would work too.

4. Just detect this error better, give a better error message that explains to notebooks that they exceeded the amount of output nbconvert can buffer.

@willingc Do you have any preferences/better suggestions on how this should be solved?

I might be able to contribute a fix but wanted some guidance/feedback before jumping into it.

**Comment by MSeal** [comment]
I strongly prefer we solve the root problem, which you suggest in # 1. I've seen a similar issue crop in rare circumstances and hadn't been able to isolate the root cause. I think captured the root issue I was seeing via your code snippet.

@djherbis If you want to help get a change we can either ship it in the 5.5 release around the corner or capture it in 5.5.1. I think this is really important to fix and can help guide development and/or testing.

Overall to fix we'll need to pull the cell timeout concept out of the polling. The wait for reply should probably be on a 5 second timeout in each loop of run_cell without failure on actual timeout (unless is_alive fails). Then if we want to enforce cell execution timeouts we could add the total time and raise if it's exceeded at the start of each loop execution.

**Comment by MSeal** [comment]
@mpacer will likely have more insight into sharp edges to consider around this loop, but we're both looking at this particular function a lot atm with 5.5 changes so I've got a mostly complete picture of the communication setup here.

**Comment by djherbis** [comment]
@MSeal Thanks, I agree with preferring # 1 :) 

I'm happy to try and contribute a fix for this, is there any contribution guideline docs etc. or should I just give a first pass at switching it over in a PR and have you two show me the way in review?

Thanks for the response!

It's starting to get late in my time zone so I'll probably dive into this tomorrow.

**Comment by MSeal** [comment]
There is a CONTRIBUTING.md file in the repo, but yes opening a PR with proposed changes is a good first step. You can label the PR with `[WIP]` if you want to ask questions and review while still working through the approach. You can reference this issue and they'll be linked.

**Comment by MSeal** [comment]
Just for future context, I did talked over the problem and discussed the various options with @mpacer .

We decided there were 3 approaches:
1) handle channel hopping with short timeouts hitting each channel at the top of the run_cell loop
2) dive into pyzmq eventloops, try to mitigate the known issues therein, probably increase some risk for rare or old setups
3) thread the channel fetching

You've started with option 1) which we think is the lowest disruption change for 5.x release cycle. It's possible we'll readdress with a pyzmq event loop in the future with >= 6.x and when python 2 is no longer supported.

Furthermore we'll still support overall cell timeout for at least 5.x releases. I'll open a separate thread after this issue is closed to discuss what makes the most sense moving forward on that topic.

**Comment by MSeal** [comment]
@djherbis I think for your commit there you'll want to do a few things.

First will be to open a Pull Request from your branch to master. You can open this from your fork page or the main code page of nbconvert. This will let us comment on the changes more easily and allow for both you and us to modify the code associated with the change.

Second, I'd change your approach slightly to make this a smaller change to the file that's easier to maintain. The different timeouts and their concerns are going to be annoying to handle just-in-time in combination with the overall timeout. I'd not make the inner loop logic care about the cell timeout and instead wait for 1 second for the exec_reply, then iterate with iopub_timeout on the iopub until that timeout is reached in total. Then at the end of the loop do the check for cell timeout. It might be a couple seconds after the actual timeout before it's noticed but that's fine.
